### ISSUE TITLE
schedules page's rail routes

### ISSUE THREAD

**Issue by albertkun** [LACMTA/metro-api-v2#435]
hello @anithegregorian! for some reason the routes for the rail are showing the old non-regional connector routes:

![image](https://github.com/LACMTA/metro-api-v2/assets/8574425/378b6134-130b-4e26-a95a-96ea7202dbbf)

Can you check to see if how you are generating those as it should be using the trip_shapes (for example https://api.metro.net/LACMTA_Rail/trip_shapes/801SB_RC_221121) looking like this:
![image](https://github.com/LACMTA/metro-api-v2/assets/8574425/07803389-2ced-4936-aafe-00100fe7f235)

**Comment by anithegregorian** [comment]
@albertkun I haven't implemented trip_shapes yet. I was looking forward to your reply on the other thread.

What I shared was on my local dev, If this API is ready then let me know. Would help immensely if I could gain some insight on how to use it, like how to get Trip ID etc.

**Comment by albertkun** [comment]
@anithegregorian ahh, ok i didn't know you were waiting for a reply! feel free to let me know what you need! for `rail` the trip id is pretty straight forward since there is just one trip_id per rail route direction:

|shape_id|route|direction
|--------|------|------|
|801NB_RC_221121| A Line|North
|801SB_RC_221121| A Line|South
|802EB_190513| B Line|East
|802WB_190513|B Line | West
|803EB_120215|C Line| East
|803WB_120215| C Line| West
|804EB_RC_221121|L Line| East
|804WB_RC_221121|L Line| West
|805EB_190513| D Line| East
|805WB_190513| D Line| West
|807NB_220810| K Line| North
|807SB_220810| K Line| South

**Comment by anithegregorian** [comment]
Hi @albertkun ,

Is it possible to include the shape in the route overview or route stops grouped API's. It would not be a good practice to define this logic in the front end (Javascript). If there's any other possibility like creating a separate endpoint then I'm fine with it as long as the logic resides on server side.

Also, how can I use shapes for bus routes?

**Comment by albertkun** [comment]
@anithegregorian for rail that might be possible since there is only one route. As a temporary solution would be able to hardcode the appropriate `trip_shapes` for the A line and E line? 

In general, just selecting one `trip_shape` for busses is difficult because there are multiple routes per bus route. We are already working on a `unique_route_stops` endpoint, but I believe that was only for the stops and not `trip_shapes`. If we can include the `trip_shapes` in that endpoint then maybe that would simplify things. 

@matikin9 is out of office today so I'll have to check in with her next week. 

But in the meantime, I'll try to see what a simplified endpoint may look like.

**Comment by anithegregorian** [comment]
@albertkun @matikin9 

I was able to get the hardcoded implementation working, do note, that it's purely for testing and demo.

While testing some route geometry doesn't match all the stops, for eg. if you check A-Line (801). And I didn't find anything for E-Line, you can check here:

https://lametro.wpengine.com/riding/schedules

On a different note if including the entire Feature/Geometry is not feasible then just the `shape_id` would also work in route overview or route stops.

Thanks, and have a great weekend!

**Comment by albertkun** [comment]
@anithegregorian Thank you for the reply! I don't think using the `shape_id` would make a difference as the `trip_shapes` is just a grouping off of the `shape_id`. 

Which endpoint are you pulling the stops from? 

Once you let me know, I'll go ahead and look into that!

I know it's getting late over there, so feel free to circle back with me on Monday! :)

**Comment by anithegregorian** [comment]
@albertkun 

Before signing off, I'm using route overview and route stops grouped API, now with one additional API call to trip shapes for Rail routes for eg. https://api.metro.net/LACMTA_Rail/trip_shapes/801NB_RC_221121

What I was referring to was the shape_id in your response https://github.com/LACMTA/metro-api-v2/issues/435#issuecomment-1887775474

It's best to get the geometry or a reference to the geometry from the previous two API calls, I don't see avoiding a third API call for geometry :)

We're off on Monday BTW following Kite Festival and Pongal...

**Comment by albertkun** [comment]
@anithegregorian ohh, that makes sense! ok, i'll try to see what makes the most sense in terms of a general shape for a line! :)

I agree if we can reduce that one extra API call that would be great, so we'll see what we can do- though it might be unavoidable since different trips and directions will have different shapes 😢 

No worries about Monday! Happy pongal and kite festival!!! 🪁 🥳

**Comment by anithegregorian** [comment]
@albertkun just checking if we can wrap this up during this week

**Comment by albertkun** [comment]
@anithegregorian yep,trying to decide what is the best way for the route_overview to reference the trip shape!

**Comment by albertkun** [comment]
@anithegregorian ok, we are going with the `route_overview` having a `shape_direction_0` and `shape_direction_1` column! however, i'm having some issues getting the automation up and running. i will be working on it over the weekend and will keep you posted!

**Comment by anithegregorian** [comment]
@albertkun noted and thanks for the quick update

**Comment by albertkun** [comment]
thank you @anithegregorian ! i appreciate your patience! 🙏

**Comment by anithegregorian** [comment]
@albertkun I can tell you're slogging this one out if you need anything from me.

**Comment by albertkun** [comment]
@anithegregorian thank you for your patience! (hopefully you haven't been seeing my late night commits, hahah..) but the endpoint should be up now!!

the geojson are stored in the route_overview endpoint (https://api.metro.net/LACMTA_Rail/route_overview) as follows:

![image](https://github.com/LACMTA/metro-api-v2/assets/8574425/f8526db4-cb30-49e9-8937-a2b1460c27e9)

the bus routes have a default geojson too, but just be aware they may not be correct depending on the day.

**Comment by albertkun** [comment]
hello @anithegregorian ! quick update, instead of the `route_overview` endpoint, we are now putting the route shapes (geojsons) into the `route_stops_grouped` endpoint:

https://api.metro.net/LACMTA_Rail/route_stops_grouped

This means to get the specific route of a line, you'd put in something like:
https://api.metro.net/LACMTA_Rail/route_stops_grouped/801

you can ignore the payload for now (which has all the stops that needs to be tweaked a bit) however, what you are looking for is within the `shape_direction_0` and `shape_direction_1` fields!

![image](https://github.com/LACMTA/metro-api-v2/assets/8574425/210d4c46-b240-4682-8d45-aabd1c5deb7b)

**Comment by anithegregorian** [comment]
Thanks @albertkun I'll check it out

**Comment by anithegregorian** [comment]
Think I found the issue, this looks like a different response, there's an extra index in the array:

![image](https://github.com/LACMTA/metro-api-v2/assets/143655/770b971f-776f-40ae-bcc2-76c67c10b22f)

**Comment by albertkun** [comment]
Ah yeah, also there's a parameter for the stop times too:

https://api.metro.net/LACMTA/trip_detail/route_code/720?geojson=true&include_stop_time_updates=false

**Comment by albertkun** [comment]
In terms of the payload field, that is being reworked, so I wouldn't recommend using it yet!

**Comment by anithegregorian** [comment]
I remember sometime back, this scenario was typical with caching (the extra index), is it so? The restructuring of APIs is a rollercoaster for both of us!

I have to use the payload to render stops, and there are a few bugs logged by Alex that need to be addressed with the next push.

**Comment by anithegregorian** [comment]
I also noticed some bus routes are not returning responses @albertkun 

https://lametro.wpengine.com/riding/schedules?line=699&agency=LACMTA

**Comment by albertkun** [comment]
> I remember sometime back, this scenario was typical with caching (the extra index), is it so? The restructuring of APIs is a rollercoaster for both of us!
> 
> I have to use the payload to render stops, and there are a few bugs logged by Alex that need to be addressed with the next push.

I haven't touched the payload part in that endpoint, so you are probably correct in remembering that it was the extra index which is being added.

**Comment by albertkun** [comment]
> I also noticed some bus routes are not returning responses @albertkun
> 
> https://lametro.wpengine.com/riding/schedules?line=699&agency=LACMTA

@matikin9 just opened the earlier issue #462 and i was able to resolve it. the problem was those routes were duplicated in the database! so i went ahead and cleared the duplicates!

**Comment by anithegregorian** [comment]
Yes @albertkun I got the updates on Bugherd. 699 should be omitted altogether.

**Comment by albertkun** [comment]
hello @anithegregorian ! hope you are doing well! just a heads up that we are re-doing the `route_stops_grouped` endpoint to organize the data by `route_code`, `direction_id`, and `day_time`, so that the client doesnt get all the schedule information all in one payload, the new endpoint will take in each of those parameters like this:

https://api.metro.net/LACMTA_Rail/route_stops_grouped/802?`day_type=weekday`&`direction_id=1`

this will return something like this:
```json
[
    {
        "route_code": "801",
        "day_type": "weekday",
        "direction_id": 0,
        "shape_direction": "value_1",
        ...
    },
    {
        "route_code": "801",
        "day_type": "weekday",
        "direction_id": 0,
        "shape_direction": "",
        ...
    },
    ...
]

The change is going up in a few hours, and I'll ping you when it is ready. In the meantime, let me know if you have any questions!

Also, if you leave the endpoint as is (https://api.metro.net/LACMTA_Rail/route_stops_grouped/802), it will return a list of the route with different `day_time`. and `direction_id`s.

**Comment by albertkun** [comment]
@anithegregorian sorry, its still taking some time... also the endpoint is going to become:

[https://api.metro.net/LACMTA_Rail/trip_shape_stop_times/802](https://api.metro.net/LACMTA_Rail/trip_shape_stop_times/802)
